x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -78,17 +79,20 @@ public void onProgress(final long progressValue) {
             assert false : end + " < " + progressValue;
             throw new IllegalArgumentException("Cannot update progress with a value greater than [end=" + end + ']');
         }
+        if (progressValue == end) {
+            return; // reached the end of the range, listeners will be completed by {@link #onResponse(Long)}
+        }
```

Comment:
Sorry for my ignorance. If a listener waits for a `progressValue` lower than the `end`, it will still be invoked before `onResponse` is called. Is this not a problem? If the listener also uses `tryRead` to access data, can it fail at `SparseFileTracker#checkAvailable` because `complete` is not updated? Or maybe it is _not_ possible for such listener to go through the `tryRead` route? In fact, I am not even sure how a listener waiting for the entire range can subsequent go through the `tryRead` route. Any pointer is appreciated!

Reply:
> Sorry for my ignorance.

No ignorance here, just good questions :)

> If a listener waits for a progressValue lower than the end, it will still be invoked before onResponse is called. Is this not a problem?

Not really a problem for the current read operation; it waited for a given range to be available because it wasn't, and it will be executed as soon as it is available. It may be an issue for a following `tryRead` (if any), which may also go on the slow read path (because `complete` may not be updated at the time it executes) and will also waits for the range to be available or will execute immediately if the range is covered by the write `progress`. Note that Lucene often uses 1kb or 4kb buffers to reads files, so Lucene can usually make 1 or more reads before having to wait a bit more.

I'm looking at ways to optimize the `complete` field update in my spacetime. Last time we talked about this it was too costly to update it on every write on all ranges, but we can maybe find a good compromise by updating for every write only if the range to be written is garanteed to make the `complete` advances (that is, all ranges before the one we write are already available).

Reply:
A couple options spring to mind:

1. Only update `complete` in assocation with firing listeners. Since if we tell nobody about the progress, there is no need to update `complete`.
2. We can check that the start of the range/gap is `>= complete` and update `complete` to progress if it is.

Reply:
Thanks Henning, those are two good ideas which I had not thought of. Worth investigating.

Reply:
@henningandersen I created the draft https://github.com/elastic/elasticsearch/pull/109247, let me know what you think 🙏🏻